### PR TITLE
✨ KCP remediates unhealthy machines

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What this PR does / why we need it**:
This PR adds to KCP support for remediating unhealthy machines according to KCP proposal changes defined by https://github.com/kubernetes-sigs/cluster-api/pull/3676

**Which issue(s) this PR fixes**:
Fixes https://github.com/kubernetes-sigs/cluster-api/issues/2976

### CODE REVIEW

**Comment by fabriziopandini** [comment]
The implementation should be completed so people can review (thanks in advance for feedback)
however, I prefer to keep this on hold while doing some additional local testing
/hold

File: api/v1alpha3/condition_consts.go
Comment (srm09) [inline api/v1alpha3/condition_consts.go:130]:
```suggestion
	// RemediationFailedReason is the reason used when a machine fails to be remediated.
```

File: controlplane/kubeadm/internal/workload_cluster_etcd.go
Comment (srm09) [inline controlplane/kubeadm/internal/workload_cluster_etcd.go:346]: Could be re-arranged this way.
```golang
var responsive bool

// if it is not possible to get an etcd member status, it is considered it not responsive.
if err != nil {
    responsive = false
} else if len(etcdStatus.Errors) > 0 {
    responsive = false 
} else {
    responsive = true
}

statuses = append(statuses, EtcdMemberStatus{
	Name:       member.Name,
	Responsive: responsive,
})
```

**Review by srm09 — COMMENTED** [review]
minor nit suggestions

File: controlplane/kubeadm/controllers/remediation.go
Comment (sedefsavas) [inline controlplane/kubeadm/controllers/remediation.go:109]: The issue in `canSafelyRemoveEtcdMember ` may get fixed during `reconcileHealth` so we need to move forward in case of error IMO. WDYT?

Comment (sedefsavas) [reply controlplane/kubeadm/controllers/remediation.go:109]: Actually, just saw that you are already doing this below by returning nil in case `!canSafelyRemediate`. 
How about here though? Not being able to connect to the workload cluster is a reason `canSafelyRemoveEtcdMember ` return error, but this check actually belongs to `reconcileHealth`.

File: controlplane/kubeadm/controllers/remediation_test.go
Comment (sedefsavas) [inline controlplane/kubeadm/controllers/remediation_test.go:23]:
```suggestion
	. "github.com/onsi/gomega"
	
```

File: controlplane/kubeadm/controllers/remediation.go
Comment (sedefsavas) [inline controlplane/kubeadm/controllers/remediation.go:77]: Given that `Spec.Replicas` defaults to 1.
```suggestion
	var desiredReplicas int = 1
```

File: controlplane/kubeadm/controllers/remediation_test.go
Comment (sedefsavas) [inline controlplane/kubeadm/controllers/remediation_test.go:48]: There is unhealthy machine here, but fails because replicas is less than 3?

File: controlplane/kubeadm/controllers/remediation.go
Comment (sedefsavas) [inline controlplane/kubeadm/controllers/remediation.go:32]: If I understand correctly, we enter here even if `MachineOwnerRemediatedCondition` is true and may set it to false below. Do we re-remediate an already remediated machine?

File: controlplane/kubeadm/controllers/controller.go
Comment (vincepri) [inline controlplane/kubeadm/controllers/controller.go:311]:
```suggestion
	if ret, err := r.reconcileUnhealthyMachines(ctx, controlPlane); !ret.IsZero() || err != nil {
```

File: controlplane/kubeadm/controllers/remediation.go
Comment (vincepri) [inline controlplane/kubeadm/controllers/remediation.go:35]:
```suggestion
	// Gets all machines that have `MachineHealthCheckSucceeded=False`,
```

File: controlplane/kubeadm/controllers/remediation.go
Comment (vincepri) [inline controlplane/kubeadm/controllers/remediation.go:36]:
```suggestion
	// (indicating a problem was detected on the machine) and `MachineOwnerRemediated` present, indicating that KCP is
```

File: controlplane/kubeadm/controllers/remediation.go
Comment (vincepri) [inline controlplane/kubeadm/controllers/remediation.go:48]: Let's track this in an issue, and schedule the work for v1alpha4?

File: controlplane/kubeadm/controllers/remediation.go
Comment (vincepri) [reply controlplane/kubeadm/controllers/remediation.go:77]: Do we even need this check? If there is defaulting in place, we should be able to assume that we can just do 
```suggestion
	desiredReplicas := *c.KCP.Spec.Replicas
```

File: controlplane/kubeadm/internal/workload_cluster_etcd.go
Comment (fabriziopandini) [reply controlplane/kubeadm/internal/workload_cluster_etcd.go:346]: reworked avoiding else branches

File: controlplane/kubeadm/controllers/remediation.go
Comment (vincepri) [inline controlplane/kubeadm/controllers/remediation.go:93]: Sidenote, it'd be great to have the conditions functions to log things out for us

File: controlplane/kubeadm/internal/machinefilters/machine_filters.go
Comment (fabriziopandini) [reply controlplane/kubeadm/internal/machinefilters/machine_filters.go:150]: MachineHealthCheckSucceeded indicates that the MHC health check is currently failing (or not), so we should check this as well.

File: controlplane/kubeadm/controllers/remediation.go
Comment (fabriziopandini) [reply controlplane/kubeadm/controllers/remediation.go:52]: We should distinguish if
- there is machine deleting but this one is the one being remediated, which is fine
- there is another machine deleting, which is blocking

File: controlplane/kubeadm/controllers/remediation.go
Comment (fabriziopandini) [reply controlplane/kubeadm/controllers/remediation.go:109]: I'm not sure I follow your comments.
ReconcileHealth does not apply any change to the cluster, so how can be it fixes an issue in  `canSafelyRemoveEtcdMember`?
Also, how do you envision to consume the output of "Not being able to connect to the workload cluster " from `reconcileHealth` in this func? And how this could prevent this func to connect to the local cluster and handling the error?

File: controlplane/kubeadm/internal/machinefilters/machine_filters.go
Comment (sedefsavas) [reply controlplane/kubeadm/internal/machinefilters/machine_filters.go:150]: I mean

```suggestion
	return conditions.IsFalse(machine, clusterv1.MachineHealthCheckSuccededCondition) && conditions.IsFalse(machine, clusterv1.MachineOwnerRemediatedCondition)
```

We don't need to continue if `MachineOwnerRemediatedCondition` is true.

File: controlplane/kubeadm/controllers/remediation.go
Comment (fabriziopandini) [reply controlplane/kubeadm/controllers/remediation.go:32]: This should not happen, because when a machine is remediated it gets deleted, and as discussed in https://github.com/kubernetes-sigs/cluster-api/pull/3830#discussion_r509631563 there are two checks to prevent double remediation.

File: controlplane/kubeadm/controllers/remediation.go
Comment (sedefsavas) [reply controlplane/kubeadm/controllers/remediation.go:52]: Okay, I thought no need to distinguish there 2 as we should wait until they are deleted and we can return early on without doing any checks after deletion is complete.

File: controlplane/kubeadm/controllers/remediation.go
Comment (sedefsavas) [reply controlplane/kubeadm/controllers/remediation.go:109]: It is not about fixing, just the correct place to return the connection error. 
`reconcileHealth ` already returns with error if it fails to connect to the workload cluster.
My thinking was not to return with errors as much as possible here if it is not directly related to remediation, but it is just a nit.

File: controlplane/kubeadm/controllers/remediation.go
Comment (fabriziopandini) [reply controlplane/kubeadm/controllers/remediation.go:48]: kk, I will open the issue. Not sure about the priority, given that this effort provides benefits only when
- two CP nodes are marked unhealthy 
- another machine has a failing etcd member

**Comment by fabriziopandini** [comment]
@vincepri @sedefsavas thanks for the feedback!
I have added some small improvements after today's tests, each of them in separate commits for helping in the review.

/hold cancel
Given the positive results of the additional local tests

File: controlplane/kubeadm/controllers/remediation.go
Comment (fabriziopandini) [reply controlplane/kubeadm/controllers/remediation.go:47]: see https://github.com/kubernetes-sigs/cluster-api/issues/3845

File: controlplane/kubeadm/controllers/remediation.go
Comment (fabriziopandini) [reply controlplane/kubeadm/controllers/remediation.go:106]: Thanks for the context.
I got your point, but from the other side we cannot prevent other func to connect to the local cluster and handling the error

**Comment by fabriziopandini** [comment]
/hold
for https://github.com/kubernetes-sigs/cluster-api/pull/3863 to merge

File: controlplane/kubeadm/controllers/remediation.go
Comment (sedefsavas) [inline controlplane/kubeadm/controllers/remediation.go:62]: nit. 
`WithOwnedConditions` is deprecated, is there another way to use here?

**Review by sedefsavas — COMMENTED** [review]
Some nits.

I will LGTM after it is approved.

**Comment by ncdc** [comment]
@sedefsavas 
> I will LGTM after it is approved.

FYI, the documented process in CONTRIBUTING.md is that LGTM is supposed to come before approval.

File: controlplane/kubeadm/internal/control_plane.go
Comment (fabriziopandini) [reply controlplane/kubeadm/internal/control_plane.go:296]: I think it is nice to have both

File: controlplane/kubeadm/controllers/remediation.go
Comment (fabriziopandini) [reply controlplane/kubeadm/controllers/remediation.go:62]: For now, I'm keeping the same approach we are using in the other controllers (WithOwnedConditions), but soon or later we should switch everything to `WithForceOverwriteConditions` unless we finally switch to server side apply.

**Comment by fabriziopandini** [comment]
/ hold
for https://github.com/kubernetes-sigs/cluster-api/pull/3900 to merge
only the latest commit is relevant for this PR

File: controllers/machinehealthcheck_controller.go
Comment (vincepri) [inline controllers/machinehealthcheck_controller.go:386]:
```suggestion
				// instead, if a remediation is in already progress, the remediation owner is responsible for completing the process and MHC should not overwrite the condition.
```

File: controlplane/kubeadm/controllers/remediation.go
Comment (vincepri) [inline controlplane/kubeadm/controllers/remediation.go:34]:
```suggestion
// reconcileUnhealthyMachines tries to remediate KubeadmControlPlane unhealthy machines based on the process described in https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/proposals/20191017-kubeadm-based-control-plane.md#remediation-using-delete-and-recreate
```

File: controlplane/kubeadm/controllers/remediation.go
Comment (vincepri) [inline controlplane/kubeadm/controllers/remediation.go:39]:
```suggestion
	// and `MachineOwnerRemediated` present, indicating that this controller is responsible for performing remediation.
```

File: controlplane/kubeadm/controllers/remediation.go
Comment (vincepri) [inline controlplane/kubeadm/controllers/remediation.go:48]:
```suggestion
	// Select the machine to be remediated, which is the oldest machine marked as unhealthy.
	//
	// NOTE: The current solution is considered acceptable for the most frequent use case (only one unhealthy machine),
```

File: controlplane/kubeadm/controllers/remediation.go
Comment (vincepri) [inline controlplane/kubeadm/controllers/remediation.go:53]:
```suggestion
	// Returns if the machine is in the process of being deleted.
```

File: controlplane/kubeadm/controllers/remediation.go
Comment (vincepri) [reply controlplane/kubeadm/controllers/remediation.go:65]: We should un-deprecate it, we shouldn't use `WithForceOverwriteConditions` here

File: controlplane/kubeadm/controllers/remediation.go
Comment (vincepri) [inline controlplane/kubeadm/controllers/remediation.go:76]:
```suggestion
	// If any of the following checks fails, we'll surface the reason in the MachineOwnerRemediated condition
```

File: controlplane/kubeadm/controllers/remediation.go
Comment (vincepri) [inline controlplane/kubeadm/controllers/remediation.go:77]:
```suggestion
```

File: controlplane/kubeadm/controllers/remediation.go
Comment (vincepri) [inline controlplane/kubeadm/controllers/remediation.go:111]:
```suggestion
```

File: controlplane/kubeadm/controllers/remediation.go
Comment (vincepri) [inline controlplane/kubeadm/controllers/remediation.go:113]:
```suggestion
			logger.Info("A control plane machine needs remediation, but removing this machine could result in etcd quorum loss. Skipping remediation", "UnhealthyMachine", machineToBeRemediated.Name)
```

File: controlplane/kubeadm/controllers/remediation.go
Comment (vincepri) [inline controlplane/kubeadm/controllers/remediation.go:125]:
```suggestion
	// If the machine that is about to be deleted is the etcd leader, move it to the newest member available.
```

File: controlplane/kubeadm/controllers/remediation.go
Comment (vincepri) [inline controlplane/kubeadm/controllers/remediation.go:148]: Hasn't remediation already happened at this point? Or is it in progress?

File: controlplane/kubeadm/controllers/remediation.go
Comment (vincepri) [inline controlplane/kubeadm/controllers/remediation.go:216]:
```suggestion
		// so we are assuming the worst scenario and considering the member unhealhy.
		//
		// NOTE: This, should not happen, given that we are running reconcileEtcdMembers before calling this method.
```

File: controlplane/kubeadm/controllers/remediation_test.go
Comment (vincepri) [inline controlplane/kubeadm/controllers/remediation_test.go:260]: Can we also add two more tests that test that this function returns false when there is only 1 or 2 machines? I know there is a guard that shouldn't let this happen, but it's good to have that test coverage

File: controlplane/kubeadm/controllers/remediation.go
Comment (fabriziopandini) [reply controlplane/kubeadm/controllers/remediation.go:148]: Yes, the machine object is deleted at this point and the KCP job is completed, but the actual machine deletion will take some time to happen, so from a user PoV I think it is correct to say that remediation is in progress (and it shows up in the same way in the conditions as well).

File: controlplane/kubeadm/controllers/remediation_test.go
Comment (fabriziopandini) [reply controlplane/kubeadm/controllers/remediation_test.go:260]: test added

**Comment by vincepri** [comment]
LGTM, squash? :)

**Review by vincepri — APPROVED** [review]
/approve
/lgtm

**Pull request merged**
